File: agent/lib/src/models/video.dart
class VideoResolution {
  final String id;
  final String videoId;
  final String resolution;
  final int width;
  final int height;
  final int bitrate;
  final String? playlistUrl;
  final int segmentsCount;
  final int totalSizeBytes;
  final String status;
  final DateTime createdAt;

  VideoResolution({
    required this.id,
    required this.videoId,
    required this.resolution,
    required this.width,
    required this.height,
    required this.bitrate,
    this.playlistUrl,
    required this.segmentsCount,
    required this.totalSizeBytes,
    required this.status,
    required this.createdAt,
  });

  factory VideoResolution.fromJson(Map<String, dynamic> json) {
    return VideoResolution(
      id: json['id'],
      videoId: json['video_id'],
      resolution: json['resolution'],
      width: json['width'],
      height: json['height'],
      bitrate: json['bitrate'],
      playlistUrl: json['playlist_url'],
      segmentsCount: json['segments_count'],
      totalSizeBytes: json['total_size_bytes'],
      status: json['status'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'video_id': videoId,
      'resolution': resolution,
      'width': width,
      'height': height,
      'bitrate': bitrate,
      'playlist_url': playlistUrl,
      'segments_count': segmentsCount,
      'total_size_bytes': totalSizeBytes,
      'status': status,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

class VideoManifest {
  final String videoId;
  final String folder;
  final String streamingMode;
  final bool watermarkEnabled;
  final String watermarkMode;
  final String watermarkColor;
  final int watermarkFontSize;
  final double watermarkOpacity;
  final int watermarkOverlayCount;
  final double watermarkInsertDuration;
  final int watermarkInsertRepeat;
  final String watermarkPosition;
  final List<VideoResolution> resolutions;
  final String defaultResolution;
  final String serverMode;
  final String downloadPolicy;
  final String modeMismatchAction;
  final VideoModePolicyInfo? videoPolicy;

  VideoManifest({
    required this.videoId,
    required this.folder,
    required this.streamingMode,
    required this.watermarkEnabled,
    required this.watermarkMode,
    required this.watermarkColor,
    required this.watermarkFontSize,
    required this.watermarkOpacity,
    required this.watermarkOverlayCount,
    required this.watermarkInsertDuration,
    required this.watermarkInsertRepeat,
    required this.watermarkPosition,
    required this.resolutions,
    required this.defaultResolution,
    this.serverMode = 'hybrid',
    this.downloadPolicy = 'allow',
    this.modeMismatchAction = 'warn',
    this.videoPolicy,
  });

  factory VideoManifest.fromJson(Map<String, dynamic> json) {
    VideoModePolicyInfo? videoPolicy;
    if (json['video_policy'] != null) {
      videoPolicy = VideoModePolicyInfo.fromJson(
        Map<String, dynamic>.from(json['video_policy']),
      );
    }

    return VideoManifest(
      videoId: json['video_id'],
      folder: json['folder'] ?? 'General',
      streamingMode: json['streaming_mode'] ?? 'hls',
      watermarkEnabled: json['watermark_enabled'] ?? true,
      watermarkMode: json['watermark_mode'] ?? 'insert',
      watermarkColor: json['watermark_color'] ?? '#FFFFFF',
      watermarkFontSize: json['watermark_font_size'] ?? 20,
      watermarkOpacity: (json['watermark_opacity'] as num?)?.toDouble() ?? 0.4,
      watermarkOverlayCount: json['watermark_overlay_count'] ?? 1,
      watermarkInsertDuration:
          (json['watermark_insert_duration'] as num?)?.toDouble() ?? 1.0,
      watermarkInsertRepeat: json['watermark_insert_repeat'] ?? 1,
      watermarkPosition: json['watermark_position'] ?? 'random',
      resolutions: (json['resolutions'] as List)
          .map(
            (res) =>
                VideoResolution.fromJson(Map<String, dynamic>.from(res as Map)),
          )
          .toList(),
      defaultResolution: json['default_resolution'] ?? '720p',
      serverMode: json['server_mode'] ?? 'hybrid',
      downloadPolicy: json['download_policy'] ?? 'allow',
      modeMismatchAction: json['mode_mismatch_action'] ?? 'warn',
      videoPolicy: videoPolicy,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'video_id': videoId,
      'folder': folder,
      'streaming_mode': streamingMode,
      'watermark_enabled': watermarkEnabled,
      'watermark_mode': watermarkMode,
      'watermark_color': watermarkColor,
      'watermark_font_size': watermarkFontSize,
      'watermark_opacity': watermarkOpacity,
      'watermark_overlay_count': watermarkOverlayCount,
      'watermark_insert_duration': watermarkInsertDuration,
      'watermark_insert_repeat': watermarkInsertRepeat,
      'watermark_position': watermarkPosition,
      'resolutions': resolutions.map((res) => res.toJson()).toList(),
      'default_resolution': defaultResolution,
      'server_mode': serverMode,
      'download_policy': downloadPolicy,
      'mode_mismatch_action': modeMismatchAction,
      'video_policy': videoPolicy?.toJson(),
    };
  }
}

class VideoModePolicyInfo {
  final String? currentAction;
  final bool banned;
  final String? modeWhenDownloaded;

  VideoModePolicyInfo({
    this.currentAction,
    required this.banned,
    this.modeWhenDownloaded,
  });

  factory VideoModePolicyInfo.fromJson(Map<String, dynamic> json) {
    return VideoModePolicyInfo(
      currentAction: json['current_action'] as String?,
      banned: json['banned'] as bool? ?? false,
      modeWhenDownloaded: json['mode_when_downloaded'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'current_action': currentAction,
      'banned': banned,
      'mode_when_downloaded': modeWhenDownloaded,
    };
  }
}

File: video-server/app/models/video.py
import uuid
from datetime import datetime
from typing import Optional
from sqlmodel import SQLModel, Field


class Video(SQLModel, table=True):
    __tablename__ = "videos"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    title: str
    description: Optional[str] = None
    original_filename: str
    original_path: str

    duration_seconds: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    codec: Optional[str] = None
    bitrate: Optional[int] = None

    status: str = "pending"
    folder: str = "General"
    streaming_mode: str = "hls"  # "hls" or "direct"

    watermark_enabled: bool = True
    watermark_mode: str = "insert"  # "insert" or "overlay"
    watermark_segments: int = 10
    watermark_text: Optional[str] = None
    watermark_color: str = "#FFFFFF"
    watermark_font_size: int = 20
    watermark_opacity: float = 0.4
    watermark_overlay_count: int = 1
    watermark_insert_duration: float = 1.0
    watermark_insert_repeat: int = 1
    watermark_position: str = "random"
    watermark_break_duration: int = 60
    encryption_key_hex: Optional[str] = None
    encryption_iv_hex: Optional[str] = None
    is_encrypted: bool = False

    storage_type: str = "local"
    storage_path: Optional[str] = None
    cdn_url: Optional[str] = None

    transcode_method: str = "local"  # "local" or "mux"
    mux_asset_id: Optional[str] = None
    mux_playback_id: Optional[str] = None

    created_by: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class VideoResolution(SQLModel, table=True):
    __tablename__ = "video_resolutions"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    video_id: str = Field(foreign_key="videos.id")
    resolution: str
    width: int
    height: int
    bitrate: int
    playlist_url: Optional[str] = None

    segments_count: int = 0
    total_size_bytes: int = 0

    status: str = "pending"

    created_at: datetime = Field(default_factory=datetime.utcnow)


class VideoSegment(SQLModel, table=True):
    __tablename__ = "video_segments"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    video_id: str = Field(foreign_key="videos.id")
    resolution_id: str = Field(foreign_key="video_resolutions.id")

    segment_number: int
    segment_hash: str
    filename: str

    duration_seconds: float
    byte_size: int

    is_watermark: bool = False
    watermark_text: Optional[str] = None

    storage_type: str = "local"
    storage_path: Optional[str] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)


class TranscodeJob(SQLModel, table=True):
    __tablename__ = "transcode_jobs"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    video_id: str = Field(foreign_key="videos.id")

    status: str = "pending"  # pending, running, completed, error, cancelled
    progress: float = 0.0

    resolutions_requested: str = "360p,480p,720p,1080p"
    resolutions_completed: str = ""

    fail_count: int = Field(default=0)
    error_message: Optional[str] = None
    priority: int = Field(default=0)

    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)


class WatchSession(SQLModel, table=True):
    __tablename__ = "watch_sessions"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    video_id: str = Field(foreign_key="videos.id")
    user_id: Optional[str] = None

    resolution: str
    segment_start: int
    segment_end: int
    completed: bool = False

    created_at: datetime = Field(default_factory=datetime.utcnow)


class CacheEntry(SQLModel, table=True):
    __tablename__ = "cache_entries"

    id: str = Field(default_factory=lambda: str(uuid.uuid4()), primary_key=True)
    cache_key: str = Field(index=True)
    cache_type: str = Field(index=True)
    file_path: Optional[str] = None
    data: Optional[str] = None
    ttl_seconds: int
    last_accessed: datetime = Field(default_factory=datetime.utcnow)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    status: str = "active"